From a6e22782281811db267e936052ec3e518ff3eb34 Mon Sep 17 00:00:00 2001
From: Pavithra Ganesan <quic_pavigane@quicinc.com>
Date: Tue, 15 Oct 2024 16:06:40 +0530
Subject: [PATCH] UPSTREAM: HOSTAPD: Send disassoc to STA only
when Link Removal Imminent field is 0 in BSS TM Request frame

When BTM req frame is sent to an ML with disassoc timer set,
it is expected to arm disassoc timer for the sta and eventually
disassociate it once the timer expires.

Current code avoids this by assuming that the ML association is
not being terminated and only the link is being removed, but link
removal is scheduled only when Link Removal Imminent field is set
in BTM as per IEEE P802.11be/D7.0 - 9.6.13.9 BSS Transition
Management Request frame format- if the Link Removal Imminent
field in the Request Mode field is equal to 1,the BSS
Termination Included field indicates that the BSS whose AP
transmits this BSS Transition Management Request frame is
shutting down, the STA associated with the AP or the non-AP
MLD that has set up only this link will be disassociated and
the non-AP MLD that has set up more than this link remains
associated to the AP MLD with the remaining setup link(s).

Hence schedule disassociation for STA unless link removal
imminent bit is set in BTM request frame

Signed-off-by: Pavithra Ganesan <quic_pavigane@quicinc.com>
---
 src/ap/wnm_ap.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/ap/wnm_ap.c
+++ b/src/ap/wnm_ap.c
@@ -1092,7 +1092,10 @@ int wnm_send_bss_tm_req(struct hostapd_data *hapd, struct sta_info *sta,
 	hapd->openwrt_stats.wnm.bss_transition_request_tx++;
 	if (disassoc_timer && sta) {
 #ifdef CONFIG_IEEE80211BE
-		if (ap_sta_is_mld(hapd, sta)) {
+		/* Link removal is scheduled only when Link Removal Imminent field is set
+		 * in BTM as per IEEE P802.11be/D7.0 - 9.6.13.9 BSS Transition
+		 * Management Request frame format else schedule full disconnection */
+		if (ap_sta_is_mld(hapd, sta) && (req_mode & WNM_BSS_TM_REQ_LINK_REMOVAL_IMMINENT)) {
 			int i;
 			unsigned int links = 0;
 
